stablehlo/dialect/VhloBytecode.cpp

```diff
@@ -696,6 +927,185 @@ void VhloBytecodeInterface::write(TypeExtensionsAttr attr,
   writer.writeSignedVarInts(attr.getBounds());
 }
 
+//===----------------------------------------------------------------------===//
+// Forked Attributes
+//===----------------------------------------------------------------------===//
+
+namespace {
+template <typename AttrOrType>
+bool assertFromVhlo(AttrOrType val) {
+  if (val.getDialect().getNamespace() != "vhlo") {
+    LLVM_DEBUG(llvm::dbgs() << "Not vhlo: " << val << '\n');
+    llvm_unreachable("All types and attributes must be VHLO for bytecode.");
+  }
+  return true;  // return a value for llvm::all_of use
+}
+}  // namespace
+
+//===----------------------------------------------------------------------===//
+// ArrayV1Attr
+
+ArrayV1Attr VhloBytecodeInterface::readArrayAttr(
+    DialectBytecodeReader &reader) const {
+  LOG_READ_CALL;
+  SmallVector<Attribute> elements;
+  if (failed(reader.readAttributes(elements))) return ArrayV1Attr();
+
+  llvm::all_of(elements, assertFromVhlo<Attribute>);
```

Comment:
It would be a `report_fatal_error` regardless of `NDEBUG` (assert is just a function name). Writing wasn't made for error reporting, much like asm printing isn't made for error reporting. Parsing/reading are.

I can flip the reader to generate errors on reading, but I still think the fatal error would be needed for printing.

Reply:
Change of plan - this is all going to move into VHLO op verifiers so we don't need assertions. I'm going to leave this ticket open until I implement that as a reminder to clean up this function.

Reply:
"I'm going to leave this ticket open". Do you mean leaving this conversation unresolved? Or you're planning to open a ticket for this?

Reply:
*Leave this conversation unresolved until it is fixed.

Will likely be back to addressing feedback on this tomorrow.

Reply:
Added verifiers to all VHLO attrs/types that hold nested attrs/types